pkg/controller/populators/forklift-populator.go

```diff
@@ -0,0 +1,593 @@
+	reconciler := &ForkliftPopulatorReconciler{
+		ReconcilerBase: ReconcilerBase{
+			client:          client,
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(forkliftPopulatorName),
+			recorder:        mgr.GetEventRecorderFor(forkliftPopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			installerLabels: installerLabels,
+		},
+		importerImage:       importerImage,
+		ovirtPopulatorImage: ovirtPopulatorImage,
+	}
+
+	forkliftPopulatorController, err := controller.New(forkliftPopulatorName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for kind, sourceType := range supportedPopulators {
+		if err := addWatchers(mgr, forkliftPopulatorController, log, kind, sourceType); err != nil {
+			return nil, err
+		}
+	}
+
+	return forkliftPopulatorController, nil
+}
+
+func addWatchers(mgr manager.Manager, c controller.Controller, log logr.Logger, sourceKind string, sourceType client.Object) error {
+	// Setup watches
+	if err := c.Watch(source.Kind(mgr.GetCache(), &corev1.PersistentVolumeClaim{}), handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			pvc := obj.(*corev1.PersistentVolumeClaim)
+			if isPVCForkliftKind(pvc) {
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+
+			// TODO(benny) check dataSourceRef PVC prime
+			if isPVCForkliftKind(pvc) {
+				owner := metav1.GetControllerOf(pvc)
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: owner.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+			return nil
+		}),
+	); err != nil {
+		return err
+	}
+
+	mapDataSourceRefToPVC := func(_ context.Context, obj client.Object) (reqs []reconcile.Request) {
+		var pvcs corev1.PersistentVolumeClaimList
+		matchingFields := client.MatchingFields{
+			dataSourceRefField: getPopulatorIndexKey(apiGroup, sourceKind, obj.GetNamespace(), obj.GetName()),
+		}
+		if err := mgr.GetClient().List(context.TODO(), &pvcs, matchingFields); err != nil {
+			log.Error(err, "Unable to list PVCs", "matchingFields", matchingFields)
+			return reqs
+		}
+		for _, pvc := range pvcs.Items {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}})
+		}
+		return reqs
+	}
+
+	if err := c.Watch(source.Kind(mgr.GetCache(), sourceType),
+		handler.EnqueueRequestsFromMapFunc(mapDataSourceRefToPVC),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile the reconcile loop for the PVC with DataSourceRef of OvirtVolumePopulator or OpenstackVolumePopulator kind
+func (r *ForkliftPopulatorReconciler) Reconcile(_ context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("PVC", req.NamespacedName)
+	log.V(1).Info("reconciling Forklift PVCs")
+	return r.reconcile(req, r, log)
```

Comment:
Why not simply pass the context passed in, to the reconcile function. In there you are creating new `context.TODO()` where you can just pass the existing context around.

Reply:
that's how it's done in all places I've seen https://github.com/kubevirt/containerized-data-importer/blob/55dd0d0110754e17af3473a2c5622dcc10c90b2f/pkg/controller/populators/populator-base.go#L278

But I've changed